### PR TITLE
[FEATURE] Json edit widget

### PR DESCRIPTION
**Pull request by domi4484**
## Description
This pr introduce a new attribute form widget to friendly display JSON data in text or tree form.
The widget can handle data from JSON or JSONB fields and it is read only.
Peek 2021-05-10 22-52

### CODE REVIEW

File: src/gui/editorwidgets/qgsjsoneditconfigdlg.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * Constructor for QgsJsonEditWidget.
+     *
+     * \param vl       The layer for which the configuration dialog will be created
+     * \param fieldIdx The index of the field on the layer for which this dialog will be created
+     * \param parent   A parent widget
+     */
+    explicit QgsJsonEditConfigDlg( QgsVectorLayer *vl, int fieldIdx, QWidget *parent = nullptr );
+
+    // QgsEditorConfigWidget interface
+  public:
```
Comment (pblottiere) [inline src/gui/editorwidgets/qgsjsoneditconfigdlg.h:48]:
```suggestion
```

File: src/gui/editorwidgets/qgsjsoneditconfigdlg.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * Constructor for QgsJsonEditWidget.
+     *
+     * \param vl       The layer for which the configuration dialog will be created
+     * \param fieldIdx The index of the field on the layer for which this dialog will be created
+     * \param parent   A parent widget
+     */
+    explicit QgsJsonEditConfigDlg( QgsVectorLayer *vl, int fieldIdx, QWidget *parent = nullptr );
+
+    // QgsEditorConfigWidget interface
```
Comment (pblottiere) [inline src/gui/editorwidgets/qgsjsoneditconfigdlg.h:47]:
```suggestion
```

File: src/gui/editorwidgets/qgsjsoneditwidgetfactory.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  public:
+
+    /**
+     * Constructor for QgsJsonEditWidgetFactory, where \a name is a human-readable
+     * name for the factory.
+     */
+    QgsJsonEditWidgetFactory( const QString &name );
+
+    // QgsEditorWidgetFactory interface
+  public:
```
Comment (pblottiere) [inline src/gui/editorwidgets/qgsjsoneditwidgetfactory.h:43]: I probably missed something but why adding a second `public` in this case too?

Comment (domi4484) [reply src/gui/editorwidgets/qgsjsoneditwidgetfactory.h:43]: It looks like most attribute form classes put inherited overridden members in a separate public block. If it is something old that we don't do anymore i can remove it.

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsjsoneditwidgetfactory.h:43]: I think you can remove it

File: src/gui/editorwidgets/qgsjsoneditwidget.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+{
+  setupUi( this );
+
+  setView( View::Text );
+
+  mCodeEditorJson->setReadOnly( true );
+  mCodeEditorJson->setHorizontalScrollBarPolicy( Qt::ScrollBarAsNeeded );
+  mCodeEditorJson->setVerticalScrollBarPolicy( Qt::ScrollBarAsNeeded );
+
+  mTreeWidget->setFont( QgsCodeEditor::getMonospaceFont() );
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/gui/editorwidgets/qgsjsoneditwidget.cpp:35]: Hm, I *think* when a stylesheet is applied (which is the case with not default fonts), it overrides the setFont(). You could add a setStylesheet( string ) to this tree widget and set the font-family to monospace. That should do it.

**Review by m-kuhn — APPROVED** [review]
I think it should be good to merge this and send it to broader testing while the fine tuning can take place soon.
Can you take care of the stylesheet next week @domi4484 ?

**Comment by domi4484** [comment]
@m-kuhn i pushed a new commit right now which sets the font via stylesheet

**Comment by roya0045** [comment]
This widget will be great to make PDAL pipelines, we just need a processing tool to make the connection.

**Comment by elpaso** [comment]
@domi4484 can you remind me why this widget is read-only?

**Comment by andreasneumann** [comment]
> @domi4484 can you remind me why this widget is read-only?

That widget was contracted with Kanton of Solothurn. We needed it only read-only. If you make it read-write, you would also need syntax check and refuse invalid JSON content. Something that we did not want to finance back then. If someone wants to step-in and finance this missing functionality of syntax-checking and refusal of invalid JSON content, then we would of course not oppose that ;-)

**Comment by elpaso** [comment]
@andreasneumann, thanks for clarifying,  I was just looking for an alternate widget for https://github.com/qgis/QGIS/issues/51911

**Pull request merged**